alf/environments/alf_gym3_wrapper.py

```diff
@@ -0,0 +1,441 @@
+        # finishes the current episode
+        self._prev_first = [False] * self.batch_size
+
+    @property
+    def batched(self):
+        return True
+
+    @property
+    def batch_size(self):
+        return self._gym3_env.num
+
+    # Implement abstract env_info_spec
+    def env_info_spec(self):
+        return self._env_info_spec
+
+    # Implement abstract observation_spec
+    def observation_spec(self):
+        return self._observation_spec
+
+    # Implement abstract action_spec
+    def action_spec(self):
+        return self._action_spec
+
+    def _create_time_step(self, reward, observation, action,
+                          step_type: List[ds.StepType]) -> ds.TimeStep:
+        """Construct a TimeStep object for Alf algorithms to consume
+
+        This function construct the TimeStep objects based on the information
+        observed from the underlying Gym3 environment. It essentially does:
+
+        1. Convert numpy arrays or lists to tensors
+        2. Apply image channel first if necessary
+        3. Trim ignored keys from the env info
+
+        """
+        observation = nest.map_structure(lambda x: torch.as_tensor(x),
+                                         observation)
+        if self._image_channel_first:
+            observation = nest.map_structure(
+                lambda x: x.permute(0, 3, 1, 2).contiguous(), observation)
+
+        trimmed_info = [
+            nest.prune_nest_like(info, self.env_info_spec())
+            for info in self._gym3_env.get_info()
+        ]
+
+        # In the case when we assume no timeouts, all episode end will
+        # be due to success or failure, where discount is set to 0.0.
+        discount = [
+            0.0 if s == ds.StepType.LAST else self._discount for s in step_type
+        ]
+
+        return ds.TimeStep(
+            step_type=torch.as_tensor(step_type),
+            reward=torch.as_tensor(reward),
+            discount=torch.as_tensor(discount),
+            observation=observation,
+            env_id=torch.arange(self.batch_size),
+            prev_action=torch.as_tensor(action),
+            env_info=nest.map_structure(
+                lambda *values: torch.as_tensor(values), *trimmed_info))
+
+    # Implement abstract _reset
+    def _reset(self) -> ds.TimeStep:
+        """Implement ``_reset()`` for compatibility
+
+        Note that by default Gym3 environmenst cannot be reset. However, if
+        ``support_force_reset`` is set to True, action -1 will be sent to all
+        the sub-environments to reset them if the underlying environment follows
+        the convention.
+
+        Otherwise the reset will be ignored.
+
+        """
+        if self._support_force_reset:
+            self._gym3_env.act(np.array([-1] * self.batch_size))
+            self._prev_first = [False] * self.batch_size
+        else:
+            logging.warning('reset() ignored by AlfGym3Wrapper')
+
+        reward, observation, _ = self._gym3_env.observe()
```

Comment:
If observation is of DictType, should we first convert it to dict here? I noticed that only observation_spec has been converted, but the observation itself it not.

Reply:
This is because I checked that `observation_spec` is `DictType`, but `observation` is actually a normal `dict`.